File: 2048.py/nc_board.py
# coding=utf-8
from turtle import Turtle
from tkMessageBox import showinfo
from tkMessageBox import showwarning
from tkMessageBox import showerror
from tkMessageBox import askyesno
import nc_cell
import random


class NCBoard:

    """ Author: LIAO
    Date: 2016.9.16
    NCLAB Copyright 2016 All right reserved
    Game Board of row * col, but for turtle, x is the horizontal direction,
    from left to right; y is the vertical direction, from bottom to up. The origin
    point is in the center of screen.
    """
    NC_NUM_CELL_ROW = 4  # Number of cells per row
    NC_NUM_CELL_COL = 4  # Number of cells per column
    NC_CELL_SIZE = 150  # Length/Width of per cell

    def __init__(self, win):

        random.seed()
        self.win = win
        self.win.bgcolor("black")
        self.win.setup(width=800, height=800)   # default size
        self.win.title("2048.py - NCLAB")

        # show help info
        showinfo('Help', u'方向键Up/Down/Left/Right\nR/r - 重新开始\nQ/q - 退出\n'
                         u'H/h - Help\n\nhttps://github.com/NCLAB2016/\n')

        # Listen Direction Key
        self.win.onkey(self.on_key_up, "Up")
        self.win.onkey(self.on_key_down, "Down")
        self.win.onkey(self.on_key_left, "Left")
        self.win.onkey(self.on_key_right, "Right")

        # Listen Game Control Key
        self.win.onkey(self.on_key_q, "Q")  # uppercase
        self.win.onkey(self.on_key_r, "R")
        self.win.onkey(self.on_key_h, "H")
        self.win.onkey(self.on_key_q, "q")  # lowercase
        self.win.onkey(self.on_key_r, "r")
        self.win.onkey(self.on_key_h, "h")

        # simple lock for control key press
        self.lock = False
        self.win.listen()

        # Draw the Board
        self.tur = Turtle()
        self.tur.color("white")
        self.tur.speed("fastest")   # draw speed
        self.tur.hideturtle()
        self.tur.pensize(3)
        self.draw_board()

        # Build Board Cells
        self.nc_cells = [[nc_cell.NCCell for col in range(self.NC_NUM_CELL_COL)] for row in range(self.NC_NUM_CELL_ROW)]
        for i in range(0, self.NC_NUM_CELL_ROW):
            for j in range(0, self.NC_NUM_CELL_COL):
                self.nc_cells[i][j] = nc_cell.NCCell((-self.NC_NUM_CELL_COL/2+j)*self.NC_CELL_SIZE,
                                                     (self.NC_NUM_CELL_ROW/2-i)*self.NC_CELL_SIZE,
                                                     i,
                                                     j,
                                                     0)

        # Random Select One Cell Set to 2/4
        self.occur_new_digit()

    def occur_new_digit(self):
        """ Occur a new digit in current board, the digit is 2 or 4,
        the position is randomly selected """
        empty_cells = []
        for i in range(0, self.NC_NUM_CELL_ROW):
            for j in range(0, self.NC_NUM_CELL_COL):
                if self.nc_cells[i][j].nc_digit == 0:
                    empty_cells.append(self.nc_cells[i][j])
        if len(empty_cells) == 0:   # no empty space
            return False
        sel = random.randint(0, len(empty_cells)-1)  # select one randomly
        empty_cells[sel].change_digit(self.tur, random.randint(1, 2) * 2)
        del empty_cells
        if self.is_game_over():
            showwarning("Thank you", "Game Over!\n")
        return True

    def is_game_over(self):
        """ Check the board state to find if game is over """
        for i in range(0, self.NC_NUM_CELL_ROW):
            for j in range(0, self.NC_NUM_CELL_COL):
                digit = self.nc_cells[i][j].nc_digit
                if digit == 0:   # exist empty place
                    return False
                if i > 0 and digit == self.nc_cells[i-1][j].nc_digit:  # check up
                    return False
                if i < self.NC_NUM_CELL_ROW - 1 and digit == self.nc_cells[i+1][j].nc_digit:    # check down
                    return False
                if j > 0 and digit == self.nc_cells[i][j-1].nc_digit:   # check left
                    return False
                if j < self.NC_NUM_CELL_COL - 1 and digit == self.nc_cells[i][j+1].nc_digit:    # check right
                    return False
        return True

    def draw_board(self):
        """ Draw the game board """
        for i in range(-self.NC_NUM_CELL_COL / 2, self.NC_NUM_CELL_COL / 2 + 1):
            # draw columns
            self.tur.penup()
            self.tur.setpos(i * self.NC_CELL_SIZE, -self.NC_NUM_CELL_ROW * self.NC_CELL_SIZE / 2)
            self.tur.pendown()
            self.tur.setheading(90)
            self.tur.forward(self.NC_NUM_CELL_ROW * self.NC_CELL_SIZE)
            # draw rows
            self.tur.penup()
            self.tur.setpos(-self.NC_NUM_CELL_COL * self.NC_CELL_SIZE / 2, i * self.NC_CELL_SIZE)
            self.tur.pendown()
            self.tur.setheading(0)
            self.tur.forward(self.NC_NUM_CELL_COL * self.NC_CELL_SIZE)

    def new_game(self):
        """ Restart a New Game """
        for i in range(0, self.NC_NUM_CELL_ROW):
            for j in range(0, self.NC_NUM_CELL_COL):
                self.nc_cells[i][j].remove_color(self.tur)
        self.occur_new_digit()

    def on_key_up(self):
        """ Press the Up Key Move All Can Move to Up """
        if self.lock:       # wait for the lock
            return
        self.lock = True    # hold the lock
        all_moved = False   # flag mark if there exists cell moved
        for col in range(0, self.NC_NUM_CELL_COL):
            while True:
                moved = False
                for i in range(0, self.NC_NUM_CELL_ROW):
                    if self.nc_cells[i][col].nc_digit != 0:
                        moved = self.moved_to_direction(i, col, 1)  # 1: up
                if not moved:
                    break
                else:
                    all_moved = True
        if all_moved:   # if exists moved, occurred a new cell randomly
            res = self.occur_new_digit()
            if not res:  # Error
                showerror("Error", "It is impossible!")
        self.lock = False   # release the lock

    def on_key_down(self):
        """ Press the Down Key Move All Can Move to Down """
        if self.lock:       # wait for the lock
            return
        self.lock = True    # hold the lock
        all_moved = False   # flag mark if there exists cell moved
        for col in range(0, self.NC_NUM_CELL_COL):
            while True:
                moved = False
                for i in range(self.NC_NUM_CELL_ROW-1, -1, -1):
                    if self.nc_cells[i][col].nc_digit != 0:
                        moved = self.moved_to_direction(i, col, 3)  # 3: down
                if not moved:
                    break
                else:
                    all_moved = True
        if all_moved:   # if exists moved, occurred a new cell randomly
            res = self.occur_new_digit()
            if not res:  # Error
                showerror("Error", "It is impossible!")
        self.lock = False

    def on_key_left(self):
        """ Press the Left Key Move All Can Move to Left """
        if self.lock:       # wait for the lock
            return
        self.lock = True    # hold the lock
        all_moved = False   # flag mark if there exists cell moved
        for row in range(0, self.NC_NUM_CELL_ROW):
            while True:
                moved = False
                for j in range(0, self.NC_NUM_CELL_COL):
                    if self.nc_cells[row][j].nc_digit != 0:
                        moved = self.moved_to_direction(row, j, 2)  # 2: left
                if not moved:
                    break
                else:
                    all_moved = True
        if all_moved:   # if exists moved, occurred a new cell randomly
            res = self.occur_new_digit()
            if not res:  # Error
                showerror("Error", "It is impossible!")
        self.lock = False

    def on_key_right(self):
        """ Press the Right Key Move All Can Move to Right """
        if self.lock:       # wait for the lock
            return
        self.lock = True    # hold the lock
        all_moved = False   # flag mark if there exists cell moved
        for row in range(0, self.NC_NUM_CELL_ROW):
            while True:
                moved = False
                for j in range(self.NC_NUM_CELL_COL-1, -1, -1):
                    if self.nc_cells[row][j].nc_digit != 0:
                        moved = self.moved_to_direction(row, j, 0)  # 0: right
                if not moved:
                    break
                else:
                    all_moved = True
        if all_moved:   # if exists moved, occurred a new cell randomly
            res = self.occur_new_digit()
            if not res:  # Error
                showerror("Error", "It is impossible!")
        self.lock = False

    def moved_to_direction(self, i, j, direction):
        """ Move cell to certain direction, merge or replace empty place """
        digit = self.nc_cells[i][j].nc_digit  # my digit

        if direction == 0:  # move right
            find_col = -1
            for col in range(j + 1, self.NC_NUM_CELL_COL):
                if self.nc_cells[i][col].nc_digit == 0:  # find empty place
                    find_col = col
                elif self.nc_cells[i][col].nc_digit == digit:  # find can merge
                    # merge digit
                    self.nc_cells[i][j].remove_color(self.tur)
                    self.nc_cells[i][col].change_digit(self.tur, digit * 2)
                    return True
                else:  # find can not merge
                    break
            if find_col == -1:  # can not find an empty place or not merge
                return False
            else:  # empty place
                self.nc_cells[i][j].remove_color(self.tur)
                self.nc_cells[i][find_col].change_digit(self.tur, digit)
            return True
        elif direction == 1:  # move up
            find_row = -1
            for row in range(i - 1, -1, -1):
                if self.nc_cells[row][j].nc_digit == 0:  # find empty place
                    find_row = row
                elif self.nc_cells[row][j].nc_digit == digit:  # find can merge
                    # merge digit
                    self.nc_cells[i][j].remove_color(self.tur)
                    self.nc_cells[row][j].change_digit(self.tur, digit * 2)
                    return True
                else:  # find can not merge
                    break
            if find_row == -1:  # can not find an empty place or not merge
                return False
            else:   # empty place
                self.nc_cells[i][j].remove_color(self.tur)
                self.nc_cells[find_row][j].change_digit(self.tur, digit)
            return True
        elif direction == 2:  # move left
            find_col = -1
            for col in range(j-1, -1, -1):
                if self.nc_cells[i][col].nc_digit == 0:  # find empty place
                    find_col = col
                elif self.nc_cells[i][col].nc_digit == digit:  # find can merge
                    # merge digit
                    self.nc_cells[i][j].remove_color(self.tur)
                    self.nc_cells[i][col].change_digit(self.tur, digit * 2)
                    return True
                else:  # find can not merge
                    break
            if find_col == -1:  # can not find an empty place or not merge
                return False
            else:  # empty place
                self.nc_cells[i][j].remove_color(self.tur)
                self.nc_cells[i][find_col].change_digit(self.tur, digit)
            return True
        elif direction == 3:  # move down
            find_row = -1
            for row in range(i + 1, self.NC_NUM_CELL_ROW):
                if self.nc_cells[row][j].nc_digit == 0:  # find empty place
                    find_row = row
                elif self.nc_cells[row][j].nc_digit == digit:  # find can merge
                    # merge digit
                    self.nc_cells[i][j].remove_color(self.tur)
                    self.nc_cells[row][j].change_digit(self.tur, digit * 2)
                    return True
                else:  # find can not merge
                    break
            if find_row == -1:  # can not find an empty place or not merge
                return False
            else:   # empty place
                self.nc_cells[i][j].remove_color(self.tur)
                self.nc_cells[find_row][j].change_digit(self.tur, digit)
            return True
        else:
            showerror("Error", "Unknown direction!")

    def on_key_q(self):
        res = askyesno("Quit", "Sure to leave me? :(")
        if res:
            self.win.bye()

    def on_key_r(self):
        res = askyesno("New", "Sure to restart? O(^_^)O")
        if res:
            self.new_game()

    def on_key_h(self):
        showinfo('Help', u'方向键Up/Down/Left/Right\nR/r - 重新开始\nQ/q - 退出\n'
                         u'H/h - Help\n\nhttps://github.com/NCLAB2016/\n')

    def draw_square(self, tur):
        for i in range(0, 4):  # Turn 4 Directions
            tur.forward(self.NC_CELL_SIZE)
            tur.left(90)

File: 2048.java/src/nclab/game/NCBoard.java
package nclab.game;

import javax.swing.*;
import java.awt.Color;
import java.awt.GridLayout;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.util.ArrayList;

/**
 * Game board which size is Height x Width
 * Created by LIAO on 2016/9/17.
 */
public class NCBoard extends JFrame {

    public static final int NC_NUM_ROW = 4; // number of rows of board
    public static final int NC_NUM_COL = 4; // number of columns of board
    public static final int NC_CELL_SIZE = 100; // size of cell
    public static final int INF = 1000000;

    // Directions
    public static final int RIGHT = 0;
    public static final int UP = 1;
    public static final int LEFT = 2;
    public static final int DOWN = 3;

    // simple lock
    public boolean lock = false;

    private NCCell[][] cells;

    public NCBoard(ImageIcon icon) {

        help();

        this.setIconImage(icon.getImage());
        this.setSize(NC_CELL_SIZE * NC_NUM_ROW, NC_CELL_SIZE * NC_NUM_COL);  // set window size
        this.setLocationRelativeTo(null);  // set position on center of screen
        this.setTitle("2048.java - NCLAB");
        this.setLayout(new GridLayout(NC_NUM_ROW, NC_NUM_COL));
        this.setResizable(false);
        this.setBackground(Color.BLACK);
        this.setDefaultCloseOperation(JFrame.DO_NOTHING_ON_CLOSE);  // set do nothing on close

        // add key listener
        this.addKeyListener(new KeyAdapter() {
            @Override
            public void keyPressed(KeyEvent e) {
                if (lock) return;
                lock = true;
                if (e.getKeyCode() == KeyEvent.VK_Q){   // quit
                    quit();
                }
                if (e.getKeyCode() == KeyEvent.VK_R){   // restart
                    restart();
                }
                else if (e.getKeyCode() == KeyEvent.VK_H) { // help
                    help();
                }
                else if (e.getKeyCode() == KeyEvent.VK_UP)  // up
                {
                    moveCells(UP);
                }
                else if (e.getKeyCode() == KeyEvent.VK_DOWN)  // down
                {
                    moveCells(DOWN);
                }
                else if (e.getKeyCode() == KeyEvent.VK_LEFT)  // left
                {
                    moveCells(LEFT);
                }
                else if (e.getKeyCode() == KeyEvent.VK_RIGHT)  // right
                {
                    moveCells(RIGHT);
                }
                lock = false;
            }
        });

        // add window listener
        this.addWindowListener(new WindowAdapter() {
            @Override
            public void windowClosing(WindowEvent e) {
                quit();
            }
        });

        // construct cells
        cells = new NCCell[NC_NUM_ROW][NC_NUM_COL];
        for (int i = 0; i < NC_NUM_ROW; ++i)
            for (int j = 0; j < NC_NUM_COL; ++j) {
                cells[i][j] = new NCCell();
                this.add(cells[i][j]);
            }

        // random select one cell to occur digit 2/4
        occurDigit();

    }

    /**
     * Show Help Message
     */
    private  void help() {
        JOptionPane.showMessageDialog(null, "Up/Down/Left/Right - Control Move\n" +
                "Q - Quit\nR - Restart\nH - Help\nhttps://github.com/NCLAB2016/",
                "Help", JOptionPane.INFORMATION_MESSAGE);
    }

    /**
     * Restart the Game
     */
    private void restart() {
        int res = JOptionPane.showConfirmDialog(null, "Sure to Restart? O(^_^)O", "Restart",
                JOptionPane.YES_NO_OPTION);
        if (res == JOptionPane.YES_OPTION) _restart();
    }

    private void _restart() {
        for (int i = 0; i < NC_NUM_ROW; ++i)
            for (int j = 0; j < NC_NUM_COL; ++j)
                cells[i][j].setDigit(0);
        occurDigit();
    }

    private void quit() {
        int res = JOptionPane.showConfirmDialog(null, "Sure to leave me? :(",
                "Quit", JOptionPane.YES_NO_OPTION);
        if (res == JOptionPane.YES_OPTION) System.exit(0);
    }

    private void moveCells(int direction) {
        boolean moved = false, allMoved = false;
        switch (direction){
            case UP:    // direction up
                while (true) {
                    moved = false;
                    for (int col = 0; col < NC_NUM_COL; ++col)
                        for (int i = 0; i < NC_NUM_ROW; ++i)
                            if (!cells[i][col].isEmpty()) {  // not empty cell
                                boolean res = movedACell(i, col, direction);
                                if (res) moved = true;
                            }
                    if (!moved) break;
                    else allMoved = true;
                }
                break;
            case DOWN:    // direction down
                while (true) {
                    moved = false;
                    for (int col = 0; col < NC_NUM_COL; ++col)
                        for (int i = NC_NUM_ROW-1; i >= 0; --i)
                            if (!cells[i][col].isEmpty()) { // not empty cell
                                boolean res = movedACell(i, col, direction);
                                if (res) moved = true;
                            }
                    if (!moved) break;
                    else allMoved = true;
                }
                break;
            case LEFT:    // direction left
                while (true) {
                    moved = false;
                    for (int row = 0; row < NC_NUM_ROW; ++row)
                        for (int j = 0; j < NC_NUM_COL; ++j)
                            if (!cells[row][j].isEmpty()) {  // not empty cell
                                boolean res = movedACell(row, j, direction);
                                if (res) moved = true;
                            }
                    if (!moved) break;
                    else allMoved = true;
                }
                break;
            case RIGHT:    // direction right
                while (true) {
                    moved = false;
                    for (int row = 0; row < NC_NUM_ROW; ++row)
                        for (int j = NC_NUM_COL-1; j >= 0; --j)
                            if (!cells[row][j].isEmpty()) { // not empty cell
                                boolean res = movedACell(row, j, direction);
                                if (res) moved = true;
                            }
                    if (!moved) break;
                    else allMoved = true;
                }
                break;
            default:    // Error
                JOptionPane.showMessageDialog(null, "Oh, it is impossible!", "Error", JOptionPane.ERROR_MESSAGE);
        }
        if (allMoved) occurDigit(); // select new cell to occur digit
        if (isGameOver()) JOptionPane.showMessageDialog(null, "Game Over!", "Thank you",
                JOptionPane.INFORMATION_MESSAGE);
    }

    private boolean isGameOver() {
        for (int i = 0; i < NC_NUM_ROW; ++i)
            for (int j = 0; j < NC_NUM_COL; ++j) {
                if (cells[i][j].isEmpty()) return false;
                int digit = cells[i][j].getDigit();
                if (i > 0 && cells[i-1][j].getDigit() == digit) return false;
                if (i < NC_NUM_ROW-1 && cells[i+1][j].getDigit() == digit) return false;
                if (j > 0 && cells[i][j-1].getDigit() == digit) return false;
                if (j < NC_NUM_COL-1 && cells[i][j+1].getDigit() == digit) return false;
            }
        return true;
    }

    private boolean movedACell(int i, int j, int dir) {
        int emptyPos = -1, digit = cells[i][j].getDigit();
        switch (dir) {
            case UP:
                for (int row = i-1; row >= 0; --row){
                    if (cells[row][j].getDigit() == digit) {  // find can merge
                        cells[i][j].setDigit(0);
                        cells[row][j].setDigit(digit << 1);   // multiply 2
                        if (cells[row][j].getDigit() >= 2048)
                            JOptionPane.showMessageDialog(null, "unbelievable! Go On!", "Ohhhh",
                                    JOptionPane.INFORMATION_MESSAGE);
                        return true;
                    }
                    if (cells[row][j].isEmpty())    // find empty cell
                    {
                        emptyPos = row;
                    }
                    else if (cells[row][j].getDigit() != digit) break;   // different
                }
                if (emptyPos == -1) return false;
                cells[i][j].setDigit(0);            // move cell
                cells[emptyPos][j].setDigit(digit);
                return true;
            case DOWN:
                for (int row = i+1; row < NC_NUM_ROW; ++row){
                    if (cells[row][j].getDigit() == digit) {  // find can merge
                        cells[i][j].setDigit(0);
                        cells[row][j].setDigit(digit << 1);   // multiply 2
                        if (cells[row][j].getDigit() >= 2048)
                            JOptionPane.showMessageDialog(null, "unbelievable! Go On!", "Ohhhh",
                                    JOptionPane.INFORMATION_MESSAGE);
                        return true;
                    }
                    if (cells[row][j].isEmpty())    // find empty cell
                    {
                        emptyPos = row;
                    }
                    else if (cells[row][j].getDigit() != digit) break; // different
                }
                if (emptyPos == -1) return false;
                cells[i][j].setDigit(0);            // move cell
                cells[emptyPos][j].setDigit(digit);
                return true;
            case LEFT:
                for (int col = j-1; col >= 0; --col){
                    if (cells[i][col].getDigit() == digit) {  // find can merge
                        cells[i][j].setDigit(0);
                        cells[i][col].setDigit(digit << 1);   // multiply 2
                        if (cells[i][col].getDigit() >= 2048)
                            JOptionPane.showMessageDialog(null, "unbelievable! Go On!", "Ohhhh",
                                    JOptionPane.INFORMATION_MESSAGE);
                        return true;
                    }
                    if (cells[i][col].isEmpty())    // find empty cell
                    {
                        emptyPos = col;
                    }
                    else if (cells[i][col].getDigit() != digit) break;   // different
                }
                if (emptyPos == -1) return false;
                cells[i][j].setDigit(0);            // move cell
                cells[i][emptyPos].setDigit(digit);
                return true;
            case RIGHT:
                for (int col = j+1; col < NC_NUM_COL; ++col){
                    if (cells[i][col].getDigit() == digit) {  // find can merge
                        cells[i][j].setDigit(0);
                        cells[i][col].setDigit(digit << 1);   // multiply 2
                        if (cells[i][col].getDigit() >= 2048)
                            JOptionPane.showMessageDialog(null, "unbelievable! Go On!", "Ohhhh",
                                    JOptionPane.INFORMATION_MESSAGE);
                        return true;
                    }
                    if (cells[i][col].isEmpty())    // find empty cell
                    {
                        emptyPos = col;
                    }
                    else if (cells[i][col].getDigit() != digit) break; // different
                }
                if (emptyPos == -1) return false;
                cells[i][j].setDigit(0);            // move cell
                cells[i][emptyPos].setDigit(digit);
                return true;
            default:    // Error
                JOptionPane.showMessageDialog(null, "Oh, it is impossible!", "Error", JOptionPane.ERROR_MESSAGE);
        }
        return false;
    }

    private void occurDigit() {

        // put all empty cell to list
        ArrayList<NCCell> list = new ArrayList<NCCell>();
        list.clear();
        for (int i = 0; i < NC_NUM_ROW; ++i)
            for (int j = 0; j < NC_NUM_COL; ++j)
                if (cells[i][j].isEmpty()) list.add(cells[i][j]);
        if (list.isEmpty())
            JOptionPane.showMessageDialog(null, "Oh, it is impossible!", "Error", JOptionPane.ERROR_MESSAGE);

        // randomly select an empty cell
        int sel = (int)(Math.random() * INF) % (list.size());
        int rndDig = Math.random() > 0.5 ? 2 : 4;

        // set select digit
        int now = -1;
        Outer: for (int i = 0; i < NC_NUM_ROW; ++i)
            for (int j = 0; j < NC_NUM_COL; ++j) {
                if (cells[i][j].isEmpty()) ++now;
                if (now == sel){
                    cells[i][j].setDigit(rndDig);
                    break Outer;
                }
            }
    }

}
